### PR TITLE
Feature/34 drop down filters

### PR DESCRIPTION
**Pull request by ColmBhandal**
Related to #34

OK, I made some progress, but there are a couple of issues. The search functionality seems to be OR-based, so adding the `+author:bodhi` for example just appends all content from Bodhi, rather than narrowing down the search. Is this what's expected?

For now, what I'm doing is just appending the filter using an "+in:" term. But it seems to just add all content of that type because the logic is OR-based so not super useful.

Maybe the logic should be AND-based?

Thanks!

### CODE REVIEW

**Comment by buddhist-uni** [comment]
Yeah, to make it "and"-based you need to add a + before all the other search terms as well... Kind of a limitation of Lunr :/

The other option is to manually filter out non-matching results in `search_index.js:displaySearchResults` by inspecting `store[results[i].ref]` to see if it matches the supplied filters.

**Comment by ColmBhandal** [comment]
@buddhist-uni 
OK. I fixed it. Seems to work OK now. Here's how it works:

1. If the category "All types" is selected, the search should work as before
1. Else, if the search term is blank, show all results in that category
1. Else, it will treat the search term as a content search, and AND that together with a search for the given category. Note: since the search term is treated as content, I believe this means it doesn't search in title etc. So the union of search results across all categories for some search term seem to be narrower than just searching for that term with "All types". I think this is OK though.

Let me know if you want any more changes, or would like to merge the PR?

**Comment by buddhist-uni** [comment]
1. Great! As it should be
2. This is okay, but a bit unnecessary as there are the content index pages that already do this. Up to you what experience you think is best.
3. You don't need to add a field specifier to use the +. You really can search for "+foo +bar" and that'll work. Further, there's a bug with this implementation, in that "+content:foo bar" will make "bar" an OR term.

I'm thinking the better implementation may be to pass "q" to the worker as normal, but also pass a "filterquery" as well. In the search index, have lunr perform both queries and then manually take the intersection of the results. What do you think of that approach?

There's also a slight UI bug here, in that if I select a category and then select All Types, it doesn't clear the results from before.

**Comment by ColmBhandal** [comment]
OK, refactored as per your Lunr suggestion. The search filter now only narrows down existing results, it doesn't display all results in a category.

Not sure about the bug but noticed this behaviour on the live site:

1. Type in 3 letters to perform a search
1. Clear those 3 letters
1. Search results persist & URL param `q` persists with old search result value
1. Clicking the search icon clears the results

The code I submitted seems to work similar now.

LMK if there are any other changes needed.

**Comment by ColmBhandal** [comment]
Note: my original attempt was to try to do it with just one Lunr call to `query` and intersect the two query conditions together... probably a better practice. But I couldn't get it to work so decided on the simpler approach of just doing two calls to `search`.

**Comment by buddhist-uni** [comment]
Okay! Now we're getting somewhere! :)

Yes, doing two calls to `search` is probably best.

Testing this out locally I noticed a couple things:
1. **History** To get reloading the page and the back and forward buttons to work properly, you'll need to add the selected type to a query parameter and to a new push/replaceState key, do a history.pushState on filter changes, replaceState when the new results come back, and make sure to properly pull and set the filter's value from the url in initial setup and from the state object in the window.onpopstate callback.  That's all a bit complicated! Happy to help if it's not clear...
2. **Non-content Types** Perhaps it would be good to also add type options for the non content types. These are searched for by `+type:foo` where `foo` is one of `pages`, `authors`, `courses`, `journals`, `posts`, `publishers`, `series`, or `tags` (where library content is of `type:content`). What do you think?

**Comment by ColmBhandal** [comment]
@buddhist-uni I've given both suggestions a crack. See my latest two commits. LMK if you need me to make any more updates, thanks :)

**Comment by buddhist-uni** [comment]
Hmmm, since the non-content and the content categories are exclusive, let's just put them together in one picker. Two is unnecessarily confusing. The JS can handle switching between the two filter types depending on which is selected (perhaps the values can include the filter type as well as the values?)

What I'm mostly looking for at this point is that this PR doesn't regress the current search behavior.

The current search supports:
1. Browser widths down to 320px (perhaps put the filter menu on a new row below the search box?)
2. The progressive loading indicator (spinner->"Loading…"->"Still loading…") reassures users on slow connections that their search is being performed (ensure this state is set whether the search is performed via the box or by changing the filter)
3. Copy and pasting the URL to a new window loads the correct results (the filter param needs to be optional, but should load the filter correctly if present)
4. The browser's back (and forward) button should go to the previous search and instantly reload its results

On my device, all I'm getting is JS errors... 😅

**Comment by buddhist-uni** [comment]
PS: There a few ways to combine the type and in fields client-side.

But another way is to simply merge them server-side in [the index generation code](https://github.com/buddhist-uni/buddhist-uni.github.io/blob/801c37fba2158ea8bc555c743959eb2939f15e32/assets/js/search_index.js):

By removing the type field:
```
"type": "{{ p.collection | default: 'pages' }}",
...
this.field('type', { boost: 0.3 });
...
'type': v.type
```

And replacing: `"category": {{ p.category | jsonify }},` 
With a unified: `"category": {{ p.category | default: p.collection | default: 'pages' | jsonify }},`

Whatever way you think is best, feel free to do. Just wanted to make sure this method was on your radar 🙂

**Comment by ColmBhandal** [comment]
I agree that they should be in one filter. I had a feeling you wanted that but I decided to just do the lazy thing and see if I'd get away with it 🤣 Will review and make updates now.

**Comment by buddhist-uni** [comment]
Okay this is looking pretty good.

My only gripe right now is that the query param is called "type" but takes a filter expression... It should either be renamed to "&filter=" or should take a simple value like "&type=av".

1. The advantage to "&filter=+type:pages" is that it's an easy change to make... 
2. The advantage to "&type=pages" is that it will allow us to add e.g. "&title=some%20phrase" in the future much more elegantly than "&filter=+title:some%20+title:phrase"

So, I'm leaning towards API 2, even though the first allows for some fun hacking (by allowing arbitrary filter expressions). Which do you think is better?

**Comment by ColmBhandal** [comment]
I agree that the second one is better for readability. But the first one has these advantages:

1. Code maintainability - we can chop and change what we're filtering and we won't need to keep doing deep code changes. It will be as simple as adding different filter expressions to that dropdown. In fact, we could even have mixed filters in the dropdown, combining a number of Lunr expressions, but we can display that to the user as something readable.
1. It's simpler for me right now.

I do agree that in this case I'd need to change the URL param to `filter`

**Comment by ColmBhandal** [comment]
BTW I hadn't fixed the type filter navigation yet when you last reviewed. I pushed a fix a few mins ago for that.

File: assets/css/search.scss [hunk 1]
```diff
@@ -6,6 +6,14 @@
     text-align: center;
 }
 
+#search-in-filter {
```
Comment (buddhist-uni) [inline assets/css/search.scss:9]: You changed the id to search-filter in the HTML and forgot to clean up the CSS.

File: search.html [hunk 2]
```diff
@@ -5,13 +5,36 @@
...
+  </div>
+  <div class="TableObject-item">
+    <button class="input-lg btn ml-2" type="submit" aria-label="Search">
+      <i title="Search" class="fas fa-search"></i>
+    </button>
+  </div>
+  <div class="TableObject-item">
+    <select id="search-filter" name="filter" class="form-select" aria-label="Filter">
+      <option value="">All types</option>
+      <option value="+in:articles">Articles</option>
```
Comment (buddhist-uni) [inline search.html:19]: Maybe alphabetical order for all the options after All types?

**Review by buddhist-uni — COMMENTED** [review]
Okay. Up to you. It's your feature! ;) Thanks for taking it on 😺

**Comment by ColmBhandal** [comment]
OK. Explanation of recent changes:

- The filter dropdown is now vertically placed below the search bar, as per your previous comment about browser width
- The CSS has been cleaned up appropriately - thanks. I'm now adding vertical padding instead of horizontal due to the new layout.
- The filter elements are alphabetically ordered - I think this is a great suggestion.

Let me know if there's anything else :)

**Review by buddhist-uni — APPROVED** [review]
Looks good! Will squash and merge.

Thanks so much for contributing 😍

**Pull request merged**